client/consensus/src/tests.rs

```diff
@@ -716,3 +290,590 @@ async fn authorities_runtime_api_tests() {
+        para_client: environ.clone().into(),
+        sync_oracle: DummyOracle,
+        para_backend: backend,
+        orchestrator_client: environ.into(),
+        orchestrator_slot_duration: SlotDuration::from_millis(SLOT_DURATION_MS),
+        orchestrator_tx_pool: orchestrator_tx_pool.clone(),
+    };
+
+    let (fut, _exit_notification_receiver) = crate::collators::lookahead::run::<
+        _,
+        Block,
+        NimbusPair,
+        _,
+        _,
+        _,
+        _,
+        _,
+        _,
+        _,
+        _,
+        _,
+        _,
+        _,
+        _,
+        _,
+    >(params);
+
+    fut.await;
+    // the block should not have been created, but the tx should be in the pool
+    assert_eq!(client.chain_info().best_number, 0);
+    assert_eq!(orchestrator_tx_pool.status().ready, 1);
+}
+
+#[tokio::test]
+async fn collate_lookahead_for_2_relay_imports_but_with_cancellation_token_stops_in_2() {
+    use substrate_test_runtime_client::DefaultTestClientBuilderExt;
+    use tokio_util::sync::CancellationToken;
+    let _ = sp_tracing::try_init_simple();
+    let keystore_path = tempfile::tempdir().expect("Creates keystore path");
+    let keystore = LocalKeystore::open(keystore_path.path(), None).expect("Creates keystore.");
+    let alice_public = keystore
+        .sr25519_generate_new(NIMBUS_KEY_ID, Some(&Keyring::Alice.to_seed()))
+        .expect("Key should be created");
+
+    // Copy of the keystore needed for tanssi_claim_slot()
+    let keystore_copy = LocalKeystore::open(keystore_path.path(), None).expect("Copies keystore.");
+    keystore_copy
+        .sr25519_generate_new(NIMBUS_KEY_ID, Some(&Keyring::Alice.to_seed()))
+        .expect("Key should be copied");
+
+    let builder = TestClientBuilder::new();
+    let backend = builder.backend();
+    let client = Arc::new(builder.build());
+    let environ = DummyFactory(client.clone());
+    // A very large number to make sure relay imports are always available
+    // the cancellation token should stop it before though
+    let relay_client = RelayChain {
+        client: client.clone(),
+        block_import_iterations: 1000u32,
+    };
+    let spawner = sp_core::testing::TaskExecutor::new();
+    let orchestrator_tx_pool = sc_transaction_pool::BasicPool::new_full(
+        Default::default(),
+        true.into(),
+        None,
+        spawner.clone(),
+        client.clone(),
+    );
+    let mock_runtime_api = MockRuntimeApi::new(Some(1000u32.into()));
+
+    let (overseer, handle) = dummy_overseer_builder(spawner.clone(), MockSupportsParachains, None)
+        .unwrap()
+        .replace_runtime_api(|_| mock_runtime_api)
+        .build()
+        .unwrap();
+    spawner.spawn("overseer", None, overseer.run().then(|_| async {}).boxed());
+
+    let cancellation_token = CancellationToken::new();
+
+    // Build the collator
+    let params = LookAheadParams {
+        create_inherent_data_providers: move |_block_hash, _| async move {
```

Comment:
If the setup code is the same for all tests, try to create a helper function. This way it will be easier to see the difference from one test to another.

Reply:
Or even better a struct like `ExtBuilder`

Reply:
will try

Reply:
Fixed [62d7ae6](https://github.com/moondance-labs/tanssi/pull/662/commits/62d7ae6e363d2c6477e92f18a88fa7b3aabdb199)